Use prime functions

spacedmonkey opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket:  https://core.trac.wordpress.org/ticket/55428
---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

spacedmonkey committed:
Use prime functions

spacedmonkey committed:
Do not prime term cache.

spacedmonkey committed:
Change variable names.

spacedmonkey committed:
Cleaner logic.

peterwilsoncc reviewed (approved):
Yes please! 

Results of manual testing:

* menu containing only post types: two queries saved for the second and subsequent post type
* menu containing terms: one query saved per term 

The `_wp_delete_post_menu_item()` and `_wp_delete_tax_menu_item()` running on the associated `delete_*` action are ensuring against a situation in which a menu item exists but the associated object does not, so this code doesn't attempt to prime object that don't exist. 

This is a very fine refactor.

spacedmonkey commented:
This PR has a number of benefits.

- For menus with multiple post types, multiple SQL requires and replaced with a simple query. 
- For menus with multiple taxonomy terms, multiple SQL requires and replaced with a simple query. 
- If posts / terms are already in memory, no query at all. 
- Those using object caching, will have caches primed. 
- Those that use object cache that support multiple gets, will get caches primed in one object cache call. 
- One less call to get_post_meta
- Much simplier code

peterwilsoncc commented on src/wp-includes/nav-menu.php:
```diff
@@ -716,49 +716,31 @@ function wp_get_nav_menu_items( $menu, $args = array() ) {
 	}
 
-	// Get all posts and terms at once to prime the caches.
-	if ( empty( $fetched[ $menu->term_id ] ) && ! wp_using_ext_object_cache() ) {
```
I don't understand why the `! wp_using_ext_object_cache()` was here. I think you're right to remove it.

If anyone has a history lesson on why it was included, I'd be interested to hear it.

spacedmonkey replied:
What this code does, if an object cache is not in use, use get_posts and get_terms to load those objects into memory. This means, one query to post / term table to get all objects in one go. This was before the prime functions existed, when this was the only
way. 

This was disabled if object caching was enabled, because the idea was, a get for single object would not be expensive. However, it still makes sense to prime caches in this. The object is in memeory and loaded in using a get multiple cache call.

spacedmonkey commented:
Merged.
